control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -299,51 +325,102 @@ func reconcileAWSEndpointService(ctx context.Context, awsEndpointService *hyperv
+			return err
+		}
+		if output == nil || output.VpcEndpoint == nil {
+			return fmt.Errorf("CreateVpcEndpointWithContext output is nil")
+		}
 
+		endpointID = *output.VpcEndpoint.VpcEndpointId
+		log.Info("endpoint created", "endpointID", endpointID)
+		awsEndpointService.Status.EndpointID = endpointID
+		endpointDNSEntries = output.VpcEndpoint.DnsEntries
 	}
 
-	if hcp.Spec.Platform.Type != hyperv1.AWSPlatform || hcp.Spec.Platform.AWS == nil || hcp.Spec.Platform.AWS.CloudProviderConfig == nil {
-		return fmt.Errorf("AWS platform information not provided in HostedControlPlane")
+	if len(endpointDNSEntries) == 0 {
+		log.Info("endpoint has no DNS entries, skipping DNS record creation", "endpointID", endpointID)
+		return nil
 	}
-	input := &ec2.CreateVpcEndpointInput{
-		ServiceName:     aws.String(serviceName),
-		VpcId:           aws.String(hcp.Spec.Platform.AWS.CloudProviderConfig.VPC),
-		VpcEndpointType: aws.String(ec2.VpcEndpointTypeInterface),
+
+	var recordName string
+	if awsEndpointService.Name == "kube-apiserver-private" {
+		recordName = "api"
+	} else if strings.HasPrefix(awsEndpointService.Name, "router-") {
+		recordName = "*.apps"
+	} else {
+		return fmt.Errorf("no mapping from AWSEndpointService to DNS")
 	}
-	if hcp.Spec.Platform.AWS.CloudProviderConfig.Subnet != nil && hcp.Spec.Platform.AWS.CloudProviderConfig.Subnet.ID != nil {
-		input.SubnetIds = []*string{hcp.Spec.Platform.AWS.CloudProviderConfig.Subnet.ID}
+
+	zoneName := fmt.Sprintf("%s.%s", hcp.Name, hypershiftLocalZone)
+	zoneID, err := lookupZoneID(ctx, route53Client, zoneName)
+	if err != nil {
+		return err
 	}
-	output, err := ec2Client.CreateVpcEndpointWithContext(ctx, input)
+
+	fqdn := fmt.Sprintf("%s.%s", recordName, zoneName)
+	err = createRecord(ctx, route53Client, zoneID, fqdn, *(endpointDNSEntries[0].DnsName))
 	if err != nil {
 		return err
 	}
+	log.Info("DNS record created", "fqdn", fqdn)
 
-	endpointID = *output.VpcEndpoint.VpcEndpointId
-	log.Info("endpoint created", "endpointID", endpointID)
-	awsEndpointService.Status.EndpointID = endpointID
+	awsEndpointService.Status.DNSName = fqdn
+	awsEndpointService.Status.DNSZoneID = zoneID
 
 	return nil
 }
 
-func (r *AWSEndpointServiceReconciler) delete(ctx context.Context, awsEndpointService *hyperv1.AWSEndpointService, ec2Client ec2iface.EC2API) (bool, error) {
+func (r *AWSEndpointServiceReconciler) delete(ctx context.Context, awsEndpointService *hyperv1.AWSEndpointService, ec2Client ec2iface.EC2API, route53Client route53iface.Route53API) (bool, error) {
 	log, err := logr.FromContext(ctx)
 	if err != nil {
 		return false, fmt.Errorf("logger not found: %w", err)
 	}
 
 	endpointID := awsEndpointService.Status.EndpointID
-	if endpointID == "" {
-		// nothing to clean up
-		return true, nil
+	if endpointID != "" {
+		if _, err := ec2Client.DeleteVpcEndpointsWithContext(ctx, &ec2.DeleteVpcEndpointsInput{
+			VpcEndpointIds: []*string{aws.String(endpointID)},
+		}); err != nil {
+			return false, err
```

Comment:
Should this handle not-found errors explicitly?

Reply:
I actually thought the same thing, then figured out that if the Endpoint doesn't exist, `DeleteVpcEndpointsWithContext` transparently returns no error :man_shrugging:

Reply:
Ha, nice. Thanks!

Reply:
For future readers of the code, wonder if it's worth throwing a comment about it here since the behavior is nonintuitive compared to other sorts of API calls? Or is this actually typical for AWS deletes and I didn't realize it?